### ISSUE TITLE
Upgrade to Rails 6.1

### ISSUE THREAD

**Issue by gravitystorm** [openstreetmap/openstreetmap-website#3021]
[Rails 6.1 was released earlier this month](https://weblog.rubyonrails.org/2020/12/9/Rails-6-1-0-release/), and will allow us to make two big changes:

* Use different buckets for different attachments. This means that we can store user images in one bucket (like now) and e.g. GPX files in another. Moving GPX files from filesystem storage (NFS) to object storage is a key item for the OSMF OWG too, to help with scaling and resilience.
* Use read-only database connections - see #2634 

Another feature in Rails 6.1 that I would like to explore is:

* Strict Loading Associations. Rails normally loads associations lazily, but that can easily lead to N+1 queries. I want to look at turning on Strict Loading, so that we need to be careful about declaring associations that should be preloaded, and reducing the number of queries that each rendering uses.

The current blockers for the upgrade is:

* Composite Primary Keys. There is a [pull request](https://github.com/composite-primary-keys/composite_primary_keys/pull/531) available, so if that gets merged we should be all set.
* I don't know of any other blockers yet!

**Comment by tomhughes** [comment]
I already have a branch for this but as you say it is currently blocked on CPK.

**Comment by tomhughes** [comment]
https://github.com/tomhughes/openstreetmap-website/tree/rails61 is my work so far...

**Comment by gravitystorm** [comment]
> https://github.com/tomhughes/openstreetmap-website/tree/rails61 is my work so far...

Great! Do the zeitwerk-related changes apply cleanly to the current codebase? If so, they could be rolled out without waiting for CPK.

**Comment by tomhughes** [comment]
They do yes - that's how I did them in fact given that the tests don't run on the rails61 branch yet.

I just hadn't got around to cherry picking them back to master but I've done it now.

**Comment by tomhughes** [comment]
So the real answer is no... There are issues in ActionMailer that only seem to appear in production mode :-(

**Comment by gravitystorm** [comment]
> So the real answer is no... There are issues in ActionMailer that only seem to appear in production mode :-(

Looks like https://github.com/rails/rails/issues/40068 - although I recommend skipping everything between the first and last comments.

> So yes, the solution is not to change inflections globally. You can afford global changes if they do not break stuff, but if they do, you need to be more precise. Fortunately, Rails 6 gives you the means to be more precise and instead of saying "ID" has to be treated that way globally, you may say
> `loader.inflector.inflect("open_id" => "OpenID")`

I've briefly tried that, by removing the global inflectors and using a zeitwerk-specific inflection list, but I got lost in a rabbit-hole around the UTF8Validator class. I'll come back to this next week unless you beat me to it!

**Comment by tomhughes** [comment]
There's a second problem in development mode with `better_errors` as well... In fact test mode is the only one that works ;-)

**Comment by tomhughes** [comment]
The development mode issue is actually a new issue caused by the recent update of the `debug_inspector` gem and specifically by https://github.com/banister/debug_inspector/pull/25.

**Comment by tomhughes** [comment]
I have now pushed eada36ff96d1b534457d98a25239ec468a7cba2c on master which appears to have been more successful...

**Comment by gravitystorm** [comment]
Yesterday I tried out the WIP branch of composite_primary_keys, and the good news is that I got it working and our tests passing (albeit with an extra patch due to an activerecord change in rails 6.1.3). 

Unfortunately it might still be a while until the next CPK release - the originally mentioned CPK PR now has conflicts with its master branch, and although someone has opened a second PR with the conflicts fixed, it doesn't seem to be making progress either. I'm not sure if there's anything that we can do to help get a CPK release out of the door, or whether we need to just wait.

**Comment by mmd-osm** [comment]
CPK pull request seems to be almost done now and ready for testing. Lots of progress during the last few hours.

**Comment by tomhughes** [comment]
Yes I'm subscribed to it so I don't really need a running commentary here ;-) Once it's released I will look to progress the work I already have in place...

**Comment by gravitystorm** [comment]
https://github.com/openstreetmap/openstreetmap-website/pull/3193

**Comment by gravitystorm** [comment]
Fixed by #3193
